envoy/server/factory_context.h

```diff
@@ -211,6 +211,13 @@ class FactoryContext : public virtual CommonFactoryContext {
    */
   virtual Stats::Scope& listenerScope() PURE;
 
+  /**
+   * @return Stats::Scope& the leaf stats scope. Factory context outlives the filter, so leafScope
+   * is the smallest scope to accommodate the per filter stats. Use listenerScope() or scope() if
+   * longer life time is needed.
+   */
+  virtual Stats::Scope& leafScope() PURE;
```

Comment:
Is there a need for access to a different scope? I haven't looked at the details, but I think it may be better to pass a different factory_context to filters, where the `scope()` returned is the correct one for that filter-chain, instead of giving them access to both and hoping they pick the correct one.

Reply:
Somewhat related: I found it was hard for *me* to figure out what the current scope is; I know what it should be, but I had to check the call sites to verify it was indeed correct. @ggreenway's suggestion sounds great in that context.

Reply:
@ggreenway 
1. scope() is currently referred to "global scope". My hunch is that it's not worth a renaming or change the semantic. 
2. Upper scopes are desired, we do have 
   a. listenerScope() called by HCM and other network filters
   b. global stats(e.g. a RDS config stat that created by a HCM config but shared by the HCMs outside this listener).

So barely one scope() might not work. Are you saying we pass two context, filter-chain-factory-context + listener-factory-context instead of a "factory-context"? That should work, at the cost that touching each network filter, maybe sub filters such as http/thrift...

Reply:
It seems like any scope accessible to the filter should have a lifetime that ends when the filterchain is removed.

I think we can leave all interfaces the same, and in `PerFilterChainFactoryContextImpl`, create new scopes for both `scope()` and `listenerScope()`. The factory will own these new scopes, and only the newly created scopes (not the parent scopes) can be accessed via that factorycontext (which is all the filter gets).

cc @jmarantz @mattklein123 can you double-check that my statements/suggestions are correct?

Reply:
That's interesting. So the global scope and listener scope life is shorter than it declared as "global" and "listener" correspondingly, am I understanding it correctly?

Reply:
IIRC scope() is actually truly global. I don't remember why this is the case but it's probably used this way for some reason. We can check the uses if we are interested in whether this can be removed, but probably not worth it.

I think there are truly per-listener stats (like listener connections, etc.), but it's conceivable this scope can just be private and not exposed in the factory. Per @ggreenway beyond this I think any scope exposed in the factory should be the per filter chain scope.

(Listener filters probably need to get the listener scope, so might need a different version of this factory).